src/routers/documents.py

```diff
@@ -0,0 +1,81 @@
+# Standard Library
+from base64 import b64encode
+from io import BytesIO
+from logging import getLogger
+from typing import Dict
+from uuid import uuid4
+
+# Third Party
+from fastapi import File, HTTPException, UploadFile
+from fastapi.routing import APIRouter
+from starlette.responses import StreamingResponse
+
+# Local
+from src.common.enums import DocumentStatus
+from src.core.process import normalize_document
+from src.database.handler import DatabaseHandler
+
+logger = getLogger(__name__)
+
+router = APIRouter()
+
+
+@router.post("/", tags=["documents"])
+async def upload_document(document: UploadFile = File(...)) -> Dict[str, str]:
+    document_id = str(uuid4())
+
+    logger.info(
+        f"({document_id}) Document named '{document.filename}' succesfully uploaded and is ready to be processed.")
+
+    # check whether the document has valid format (PDF)
+    if document.content_type != 'application/pdf':
+        logger.debug(f"({document_id}) Invalid document content type '{document.filename}'.")
+        raise HTTPException(
+            status_code=415,
+            detail=f"Invalid document content type '{document.content_type}', only 'application/pdf' is supported.")
+
+    # save header of the document to the database
+    DatabaseHandler().insert_document(document_id=document_id, document_name=document.filename)
+
+    # read and encode the content of the document
+    document_content = await document.read()
+    document_content_encoded = b64encode(document_content).decode('ascii')
+
+    # offload the document processing
+    normalize_document.send(document_id=document_id,
```

Comment:
I agree, this was something that I was thinking about whether I should send it thru the queue or save it to the db directly. I did not have any experience with dramatiq before, I thought that it might be faster to send it than save it. Related to the bullet 1, I am more familiar with Kafka, however when I check https://www.rabbitmq.com/tutorials/tutorial-two-dotnet.html -> Message durability, you can actually set that messages are durable and persisted on the disk, so they should survive the crash or restart. There is still a chance that the message could be lost but so when saving to db.

Reply:
Makes sense :)